src/ai/backend/manager/cleaner.py

```diff
@@ -0,0 +1,160 @@
+    heuristic_interval_weight: float = 0.4,  # NOTE: to repeat more than twice within the same time window.
+) -> float:
+    if isinstance(threshold, relativedelta):  # months, years
+        return max_interval
+    return min(max_interval, threshold.total_seconds() * heuristic_interval_weight)
+
+
+async def collect_and_terminate(
+    db: ExtendedAsyncSAEngine,
+    lifecycle_manager: SessionDestroyer,
+    status: SessionStatus | KernelStatus,
+    threshold: TimeDelta,
+    /,
+    interval: float,
+):
+    now = datetime.now(tz=tzutc())
+    query = (
+        sa.select(SessionRow)
+        .join(KernelRow, KernelRow.session_id == SessionRow.id)
+        .where(
+            or_(
+                # Condition for session-level
+                and_(
+                    SessionRow.status == status,
+                    (
+                        now
+                        - SessionRow.status_history[status.name].astext.cast(
+                            sa.types.DateTime(timezone=True)
+                        )
+                        >= threshold
+                    ),
+                ),
+                # Condition for kernel-level
+                and_(
+                    KernelRow.status == status,
+                    (
+                        now
+                        - KernelRow.status_history[status.name].astext.cast(
+                            sa.types.DateTime(timezone=True)
+                        )
+                        >= threshold
+                    ),
+                ),
+            )
+        )
+        .distinct(SessionRow.id)
+        .options(
+            noload("*"),
+            load_only(SessionRow.id, SessionRow.name, SessionRow.access_key),
+        )
+    )
+
+    try:
+        async with db.begin_readonly() as conn:
+            result = await conn.execute(query)
+            sessions = result.fetchall()
+    except Exception as e:
+        log.error(e)
+        raise e
+
+    results_and_exceptions = await asyncio.gather(
+        *[
+            asyncio.create_task(
+                lifecycle_manager.destroy_session(
+                    session, forced=True, reason=KernelLifecycleEventReason.HANG_TIMEOUT
+                ),
+            )
+            for session in sessions
+        ],
+        return_exceptions=True,
+    )
+    for result_or_exception in results_and_exceptions:
+        if isinstance(result_or_exception, (BaseException, Exception)):
+            log.error(
+                "hanging session force-termination error: {}",
+                repr(result_or_exception),
+                exc_info=result_or_exception,
+            )
+
+
+@actxmgr
+async def hanging_session_scanner_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+    try:
+        session_hang_tolerance = session_hang_tolerance_iv.check(
+            await root_ctx.shared_config.etcd.get_prefix_dict("config/session/hang-tolerance")
+        )
+    except Exception as e:
+        log.error(e)
+        raise e
```

Comment:
Why not read all configurations initially on config read?

Reply:
To follow the pattern below:
https://github.com/lablup/backend.ai/blob/d9b1ef1b6c2169455b0833f97b3b072409ac8aa3/src/ai/backend/manager/server.py#L916-L940